src/main/java/org/reactome/addlinks/referencecreators/EnsemblReferenceCreator.java

```diff
@@ -0,0 +1,115 @@
+     * @param sourceReferences - A list of GKInstance objects that were in the database. References will be created if they are in the keys of mapping and also in sourceReferences.
+     * This is to handle cases where a mapping comes from a third-party file that may contain source identifiers that are in *their* system but aren't actually in Reactome.
+     * @throws Exception
+     */
+    @Override
+    public void createIdentifiers(long personID, Map<String, Map<String, List<String>>> mappings, List<GKInstance> sourceReferences) throws Exception {
+
+        // Iterate through each ReferenceGeneProduct (RGP) instance in the DB.
+        logger.info("Creating Ensembl cross-references");
+        for (GKInstance sourceInst : sourceReferences) {
+            String sourceIdentifier = (String) sourceInst.getAttributeValue(ReactomeJavaConstants.identifier);
+            // Determine which species the instance pertains too. Retrieve corresponding mapping structures.
+            GKInstance speciesInst = (GKInstance) sourceInst.getAttributeValue(ReactomeJavaConstants.species);
+            String biomartSpeciesName = getBiomartSpeciesName(speciesInst.getDisplayName());
+            // Iterate through each Ensembl Protein identifier. Retrieve any corresponding Transcript and Gene
+            // identifiers and add each to a set. These will be used to create cross references to Ensembl from Reactome.
+            Set<String> ensemblIds = collectEnsemblIdentifiers(mappings, sourceIdentifier, biomartSpeciesName);
+            // Iterate through each Ensembl identifier and create a ? instance.
+            for (String ensemblId : ensemblIds) {
+                if (!this.checkXRefExists(sourceInst, ensemblId))
+                {
+                    if (!this.testMode)
+                    {
+                        this.refCreator.createIdentifier(ensemblId, String.valueOf(sourceInst.getDBID()), targetRefDB, personID, this.getClass().getName(), speciesInst.getDBID());
+                    }
+                }
+            }
+        }
+    }
+
+    /**
+     * Retrieve all Ensembl Gene, Transcript and Protein identifiers associated with the source instance identifier.
+     * @param mappings Large mapping structure generated in FileProcessor step.
+     * @param sourceIdentifier Identifier taken from RGP instance.
+     * @param biomartSpeciesName Species name in biomart format (eg: hsapiens).
+     * @return Filtered Set of Ensembl identifiers (Gene, Transcript and Protein) associated with RGP identifier.
+     */
+    private Set<String> collectEnsemblIdentifiers(Map<String, Map<String, List<String>>> mappings, String sourceIdentifier, String biomartSpeciesName) {
+        String proteinToTranscriptsKey = biomartSpeciesName + "_proteinToTranscripts";
+        String proteinToGenesKey = biomartSpeciesName + "_proteinToGenes";
+        String uniprotToENSPKey = biomartSpeciesName + "_uniprotToProteins";
```

Comment:
These suffixes get used in a few places, right? Might make sense to create some String constants for them, and put them in a place where any class that needs them can access them.